### PR TITLE
Add visualDensity to checkbox, radio, icon button, and chip.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description
This adds a visualDensity attribute to checkbox, radio, icon button, and chip.

## Tests
- added tests for all components.
## Breaking Change
- [X] No, this is *not* a breaking change.

### CODE REVIEW

File: dev/manual_tests/lib/density.dart [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+              Padding(
+                padding: const EdgeInsets.all(8.0),
+                child: Row(
+                  children: <Widget>[
+                    const Text('Text Scale'),
+                    Expanded(
+                      child: SliderTheme(
+                        data: controlTheme,
+                        child: Slider(
+                            label: '${widget.model.size}',
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]: for entire code block: 
```suggestion
                          label: '${widget.model.size}',
```

File: dev/manual_tests/lib/density.dart [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+                            min: 0.5,
+                            max: 3.0,
+                            onChanged: (double value) {
+                              widget.model.size = value;
+                            },
+                            value: widget.model.size),
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]:
```suggestion
                            value: widget.model.size,
                          ),
```

File: dev/manual_tests/lib/density.dart [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+  }
+
+  void _modelChanged() {
+    setState(() {});
+  }
+
+  double sliderValue = 0.0;
+
+  String _densityToProfile(VisualDensity density) {
+    if (density == VisualDensity.standard) {
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]: Looks cleaner to use if/else statements here

File: dev/manual_tests/lib/density.dart [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+              Padding(
+                padding: const EdgeInsets.all(8.0),
+                child: Row(
+                  children: <Widget>[
+                    const Text('X Density'),
+                    Expanded(
+                      child: SliderTheme(
+                        data: controlTheme,
+                        child: Slider(
+                            label: '${widget.model.density.horizontal.toStringAsFixed(1)}',
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]: whole code block:
```suggestion
                          label: '${widget.model.density.horizontal.toStringAsFixed(1)}',
```

File: dev/manual_tests/lib/density.dart [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+                            min: VisualDensity.minimumDensity,
+                            max: VisualDensity.maximumDensity,
+                            onChanged: (double value) {
+                              widget.model.density = widget.model.density.copyWith(horizontal: value, vertical: widget.model.density.vertical);
+                            },
+                            value: widget.model.density.horizontal),
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]:
```suggestion
                            value: widget.model.density.horizontal,
                          ),
```

File: dev/manual_tests/lib/density.dart [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+              Padding(
+                padding: const EdgeInsets.all(8.0),
+                child: Row(
+                  children: <Widget>[
+                    const Text('Y Density'),
+                    Expanded(
+                      child: SliderTheme(
+                        data: controlTheme,
+                        child: Slider(
+                            label: '${widget.model.density.vertical.toStringAsFixed(1)}',
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]:
```suggestion
                          label: '${widget.model.density.vertical.toStringAsFixed(1)}',
```

File: dev/manual_tests/lib/density.dart [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+                            min: VisualDensity.minimumDensity,
+                            max: VisualDensity.maximumDensity,
+                            onChanged: (double value) {
+                              widget.model.density = widget.model.density.copyWith(horizontal: widget.model.density.horizontal, vertical: value);
+                            },
+                            value: widget.model.density.vertical),
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]:
```suggestion
                            value: widget.model.density.vertical,
                          ),
```

File: dev/manual_tests/lib/density.dart [hunk 8]
```diff
@@ -0,0 +1,572 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (shihaohong) [reply dev/manual_tests/lib/density.dart]: Also, how do manual tests work? Once they're created here, is the idea that someone could just come in and poke and prod at the app?

[the remark this answers is not in the document]
Comment (shihaohong) [reply dev/manual_tests/lib/density.dart]: I started reviewing this and realized that this file is already in `master` but a rebase hasn't been performed yet. I'll leave the comments I've already made in here anyway

File: packages/flutter/test/material/chip_test.dart [hunk 9]
```diff
@@ -2245,6 +2245,84 @@ void main() {
...
+                ],
+              ),
+            ),
+          ),
+        ),
+      );
+    }
+
+    await buildTest(const VisualDensity());
+    RenderBox box = tester.renderObject(find.byKey(key));
```
Comment (shihaohong) [inline packages/flutter/test/material/chip_test.dart]: Analyzer is suggesting:
```suggestion
    final RenderBox box = tester.renderObject(find.byKey(key));
```

File: packages/flutter/test/material/icon_button_test.dart [hunk 10]
```diff
@@ -467,6 +467,43 @@ void main() {
...
+    Future<void> buildTest(VisualDensity visualDensity) async {
+      return await tester.pumpWidget(
+        MaterialApp(
+          home: Material(
+            child: Center(
+              child: IconButton(
+                visualDensity: visualDensity,
+                key: key,
+                onPressed: () {},
+                icon: Icon(Icons.play_arrow),
```
Comment (shihaohong) [inline packages/flutter/test/material/icon_button_test.dart]:
```suggestion
                icon: const Icon(Icons.play_arrow),
```

File: packages/flutter/test/material/chip_test.dart [hunk 11]
```diff
@@ -2245,6 +2245,84 @@ void main() {
...
+    Rect textBox = tester.getRect(find.byKey(textKey));
+    Rect iconBox = tester.getRect(find.byKey(iconKey));
+    Rect avatarBox = tester.getRect(find.byKey(avatarKey));
+    await tester.pumpAndSettle();
+    expect(box.size, equals(const Size(128, 48)));
+    expect(textBox, equals(const Rect.fromLTRB(372, 17, 428, 31)));
```
[the remark this answers is not in the document]
Comment (shihaohong) [reply packages/flutter/test/material/chip_test.dart]: That makes sense, and yes, I think this will be easier to understand. Thank you!

File: packages/flutter/test/material/flat_button_test.dart [hunk 12]
```diff
@@ -705,7 +704,7 @@ void main() {
                 visualDensity: visualDensity,
                 key: key,
                 onPressed: () {},
-                child: useText ? const Text('Text', key: childKey) : Container(key: childKey, width: 100, height: 100, color: const Color(0xffff0000)),
+                child: useText ? const Text('Text') : Container(width: 100, height: 100, color: const Color(0xffff0000)),
```
Comment (shihaohong) [inline packages/flutter/test/material/flat_button_test.dart]: I thought I left a comment here in my previous review, but was this change intentional?

Comment (gspencergoog) [reply packages/flutter/test/material/flat_button_test.dart]: And I thought I already reverted those changes.  It's just the result of a bad merge.  I think I've reverted them now.

File: dev/manual_tests/lib/density.dart [hunk 13]
```diff
@@ -458,6 +481,86 @@ class _MyHomePageState extends State<MyHomePage> {
...
+                      });
+                    }
+                  : null,
+              groupValue: radioValue,
+              value: index,
+            );
+          }),
+        ),
+      ),
+      _ControlTile(
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]: IconButton is duplicated above

File: dev/manual_tests/lib/density.dart [hunk 14]
```diff
@@ -458,6 +481,86 @@ class _MyHomePageState extends State<MyHomePage> {
...
+          mainAxisAlignment: MainAxisAlignment.center,
+          children: List<Widget>.generate(iconValues.length, (int index) {
+            return IconButton(
+              onPressed: _model.enable ? () {} : null,
+              icon: Icon(iconValues[index]),
+            );
+          }),
+        ),
+      ),
+      _ControlTile(
```
Comment (shihaohong) [inline dev/manual_tests/lib/density.dart]: So is RadioButton

File: dev/manual_tests/lib/density.dart (hunk 13 above)
Comment (gspencergoog) [reply dev/manual_tests/lib/density.dart]: Argh.  Removed!

File: dev/manual_tests/lib/density.dart (hunk 14 above)
Comment (gspencergoog) [reply dev/manual_tests/lib/density.dart]: Removed

**Review by shihaohong — APPROVED** [review]
LGTM
